src/approx/terrain.py

```diff
@@ -245,194 +245,203 @@ def build_ls_matrix(u_knots, v_knots, terrain, sparse=False):
     v_n_basf = len(v_knots) - 3
     terrain_len = len(terrain)
 
+    row = numpy.zeros(terrain_len * 9 )
+    col = numpy.zeros(terrain_len * 9 )
+    data = numpy.zeros(terrain_len * 9 )
+
     if sparse is True:
-        mat_b = scipy.sparse.dok_matrix((terrain_len, u_n_basf * v_n_basf))
-        interval = numpy.empty((terrain_len, 2))
+        nnz_b = 0
     else:
         mat_b = numpy.empty((terrain_len, u_n_basf * v_n_basf))
-        interval = numpy.empty((terrain_len, 2))
+
+    interval = numpy.empty((terrain_len, 2))
 
     for idx in range(0, terrain_len):
-        u_base_vec, i_idx = spline_base_vec(u_knots, terrain[idx, 0], sparse)
-        v_base_vec, j_idx = spline_base_vec(u_knots, terrain[idx, 1], sparse)
+        u_base_vec, i_idx = spline_base_vec(u_knots, terrain[idx, 0], 0, sparse)
+        v_base_vec, j_idx = spline_base_vec(v_knots, terrain[idx, 1], 0, sparse)
         if sparse is True:
-            # possible problem
-            mat = scipy.sparse.kron(u_base_vec.transpose(), v_base_vec.transpose(), "dok")
-            mat_b[idx, :] = mat.transpose()
+            # Hard-coded Kronecker product (problem based)
+            for n in range(0,3):
+                data[nnz_b + 3 * n:nnz_b + 3 * (n + 1)] = v_base_vec[n] * u_base_vec
+                for m in range(0,3):
+                    col[nnz_b + (3 * n) + m] = (j_idx + n) * u_n_basf + i_idx + m
+            row[nnz_b:nnz_b+9] = idx
+            nnz_b +=  9
         else:
-            # possible problem
-            mat_b[idx] = numpy.kron(u_base_vec.transpose(), v_base_vec.transpose())
+            mat_b[idx] = numpy.kron(v_base_vec, u_base_vec)
+
         interval[idx][0] = i_idx
         interval[idx][1] = j_idx
 
+    if sparse is True:
+        mat_b = scipy.sparse.csr_matrix((data,(row,col)),shape=(terrain_len, u_n_basf * v_n_basf))
+
     return mat_b, interval
 
```

Comment:
Following function contains not descriptive names of variables. Try to change it from eg "u" to "u_param" and "a" to "vec_a"